src/v/datalake/datalake_manager.cc

```diff
@@ -264,128 +274,93 @@ ss::future<> datalake_manager::start() {
-    return std::max(min_translation_interval, _iceberg_commit_interval() / 3);
-}
-
-void datalake_manager::on_group_notification(const model::ntp& ntp) {
-    auto partition = _partition_mgr->local().get(ntp);
-    if (!partition || !model::is_user_topic(ntp)) {
-        return;
+ss::future<>
+datalake_manager::handle_translator_state_change(const model::ntp& ntp) {
+    if (_gate.is_closed() || !model::is_user_topic(ntp)) {
+        co_return;
     }
+    auto partition = _partition_mgr->local().get(ntp);
+    auto is_leader = partition && partition->raft()->is_leader();
     const auto& topic_cfg = _topic_table->local().get_topic_cfg(
       model::topic_namespace_view{ntp});
-    if (!topic_cfg) {
-        return;
-    }
-    auto it = _translators.find(ntp);
-    // todo(iceberg) handle topic / partition disabling
-    auto iceberg_disabled = topic_cfg->properties.iceberg_mode
-                            == model::iceberg_mode::disabled;
-    if (!partition->is_leader() || iceberg_disabled) {
-        if (it != _translators.end()) {
-            stop_translator(partition->ntp());
-        }
-        return;
-    }
-
-    // By now we know the partition is a leader and iceberg is enabled, so
-    // there has to be a translator, spin one up if it doesn't already exist.
-    if (it == _translators.end()) {
-        start_translator(
-          partition,
-          topic_cfg->properties.iceberg_mode,
-          topic_cfg->properties.iceberg_invalid_record_action.value_or(
-            config::shard_local_cfg().iceberg_invalid_record_action.value()));
-    } else {
-        // check if translation interval changed.
-        auto target_interval = translation_interval_ms();
-        if (it->second->translation_interval() != target_interval) {
-            it->second->reset_translation_interval(target_interval);
-        }
-
-        // check if invalid record action changed.
-        auto target_action
-          = topic_cfg->properties.iceberg_invalid_record_action.value_or(
-            config::shard_local_cfg().iceberg_invalid_record_action.value());
-
-        vlog(
-          datalake_log.warn,
-          "Invalid record action: {} vs {}",
-          target_action,
-          it->second->invalid_record_action());
-
-        if (it->second->invalid_record_action() != target_action) {
-            it->second->reset_invalid_record_action(target_action);
+    const auto& translators = _scheduler.all_translators();
+    auto translator_it = translators.find(ntp);
+    auto translator_exists = translator_it != translators.end();
+    auto iceberg_disabled = topic_cfg
+                            && topic_cfg->properties.iceberg_mode
+                                 == model::iceberg_mode::disabled;
+    auto requires_active_translator = partition && topic_cfg
+                                      && !iceberg_disabled && is_leader;
+    if (translator_exists && !requires_active_translator) {
```

Comment:
nit: maybe just me, but i think these conditions would be easier followed as

```
if !requires_active_translator:
  if translator_exists:
    remove translator
  return

if translator_exists:
  reconcile properties
  return

...set up translator
```

Reply:
yeah i tend to agree. i've taken a light touch on the existing code so far, but happy to do this type of cleanup.

Reply:
done